### DISCUSSION TITLE
How to have new Vocab and its Training

### DISCUSSION THREAD

**Discussion by osman-aktepe** [mindee/doctr#608] [Q&A] [answered]
Hi all,

First of all, I tested and admired your project. What I want is to add turkish vocab, and retrain the recognition model.

My question is;

Can I use classification training under references section? Or do i have to use recognition training with prepared data. Also If your answer is recognition part, do you have any data generator from texts for that format? 

I could not find a document and, in the discussion part, I could not see any explanation about it. If there is, please guide me and forgive me.

Regards

**Comment by charlesmindee** [comment, marked as answer]
Hi @osman-aktepe, thank you for your interest in docTR!

If you want to train with turkish vocab, we first need to integrate this vocab to doctr indeed. Then, you need to retrain a recognition model with this vocab on a turkish dataset (images of word boxes + corresponding annotations).
If you don't have such a dataset, you can either collect turkish pictures of words and then annotate them manually (or pass them through another OCR to get them annotated), or you can generate a fully synthetic dataset writing words on images with different fonts, sizes, colors, ... and you have directly the annotations because you know the words you just drew.

I hope this answer your question ! :smile:

**Reply by osman-aktepe** [reply]
Hi @charlesmindee,

thanks for your answer, so basically you are forwarding me for recognition part.
I can add the turkish vocab by cloning your repo, if i need a quick test which is ok. I have read source codes last night :)
I do not have that kind of dataset and It does not make sense to collect and prepare all of those images. I can write some apps to generate those. 

I have two differente questions;
- What is character classification for in this project? I could not find any direct link.
- How many images should i create and what should be their max min sizes as letter count and pixels. Because I tested your pretrained models for english and they work better for text which has bigger sizes. I want to have close success ratio so guiding would be amazing :)

Thanks

**Reply by charlesmindee** [reply]
In docTR, character classification is a MNIST-like task, where we generate squared character pictures and we train a backbone to classify the characters of the selected vocab (instead of classifying digits).
The recognition model typically takes 32 x 128 as input size (aspect ratio of 4), and we pad the images which have a different aspect ratio with 0 (either shorter or longer words) to this size. You can generate words between 1 and 25 characters for instance, it should work fine, with a font size large enough to occupy most of the available space on the 32 x 128 canvas.

**Reply by osman-aktepe** [reply]
Hi @charlesmindee ,

I want to clarify several things too,

- Is it enough to add extra line for turkish vocab here: https://github.com/mindee/doctr/blob/b27c3a664e99989e11d7bc527278abc33a539db0/doctr/datasets/vocabs.py or do I need to do something more?
- For example 500-600 images are enough or should I create more?
- Should I create words with punctuations?
Thanks

**Reply by charlesmindee** [reply]
Hi @osman-aktepe, 
- yes it is enough to add the extra line there (with a PR)
- I think 500-600 word images is too few, we use several millions on our side but maybe a few thousands/tens of thousands/hundreds of thousands are enough. It depends on the maximum word length you want your model to be robust on, if you want it to be able to predict 20-chars words with a lot of different chars in your vocab, you really need a lot of data.
- If you want you model to learn punctuation, you need to insert punctuations in your words, for instance: "home:", "coffee.", "..."

**Reply by osman-aktepe** [reply]
Hi @charlesmindee,

I know I am bothering you but now I am coding for synthetic dataset generator and I want to ask some questions. I am trying to follow your guidance and I do not want to do something ridiculous :)

- Should the overall size be 32 x 128? For example should i fit 20 chars on 32 x 128?
- If I create a big image, will it be resized or splitted?  
- Do you have any small dataset file to download to understand what kind of backgrounds and image kinds should I create?

Thanks for your help

**Reply by charlesmindee** [reply]
Hi @osman-aktepe ,
This is a sample of our dataset:
[images.zip](https://github.com/mindee/doctr/files/7555006/images.zip)
Then we resize each image to 32 x 128 (input size of the recognition model), when we feed the model. We have a preprocessing pipe which performs resizing, normalization, standardization, image augmentations, batching. You can either train with our script and our pipeline, or create your own preprocessing pipeline. :smile: 
I hope this answer your question!

**Reply by osman-aktepe** [reply]
Hi @charlesmindee ,

This explains everytihng :) I will use your pipeline.

Thanks a lot
